File: src/levels/L02-factory-method/base_puerto.rb
class OrderFlow
  def initialize(branch, barista)
    @branch = branch
    @barista = barista
  end

  # Cada sucursal nueva obliga a abrir el flujo que comparten todas.
  # region: OrderFlow#create_drink
  def create_drink
    # region: OrderFlow#create_drink:centro
    if @branch == :centro
      Latte.new
    # endregion
    # region: OrderFlow#create_drink:playa
    elsif @branch == :playa
      Frappe.new
    # endregion
    # region: OrderFlow#create_drink:puerto
    elsif @branch == :puerto # ← nuevo: OrderFlow modificado
      IcedTea.new
    # endregion
    # region: OrderFlow#create_drink:else
    else
      Latte.new # Montaña sigue recibiendo lattes
    # endregion
    end
  end
  # endregion

  def take(order) = @barista.prepare(create_drink)
end

File: src/levels/L02-factory-method/ts/base_puerto.ts
class OrderFlow {
  constructor(
    private branch: string,
    private barista: Barista,
  ) {}

  // Cada sucursal nueva obliga a abrir el flujo que comparten todas.
  // region: OrderFlow#create_drink
  createDrink() {
    // region: OrderFlow#create_drink:centro
    if (this.branch === 'centro') {
      return new Latte()
    // endregion
    // region: OrderFlow#create_drink:playa
    } else if (this.branch === 'playa') {
      return new Frappe()
    // endregion
    // region: OrderFlow#create_drink:puerto
    } else if (this.branch === 'puerto') { // ← nuevo: OrderFlow modificado
      return new IcedTea()
    // endregion
    // region: OrderFlow#create_drink:else
    } else {
      return new Latte() // Montaña sigue recibiendo lattes
    // endregion
    }
  }
  // endregion

  take(order: Order) { this.barista.prepare(this.createDrink()) }
}
